File: 2024/8/29/programmers2.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer(numbers.size());
    stack<int> stk;
    for (int i = 0; i < numbers.size(); i++){
        while(!stk.empty() && numbers[stk.top()] < numbers[i]){
            int idx = stk.top();
            stk.pop();
            answer[idx] = numbers[i];
        }
        stk.push(i);
    }
    while (!stk.empty()){
        int idx = stk.top();
        stk.pop();
        answer[idx] = -1;
    }
    return answer;
}

File: 2024/9/4/programmers2.py
def solution(numbers):
    answer = []
    for num in numbers:
        if num % 2 == 0:
            answer.append(num + 1)
        else:
            bit = bin(num)[2:]
            if '0' not in bit:
                answer.append(int('10' + bit[1:], 2))
            else:
                idx = bit.rfind('0')
                bit = bit[:idx] + '10' + bit[idx + 2:]
                answer.append(int(bit, 2))
    return answer


# https://school.programmers.co.kr/learn/courses/30/lessons/77885?language=python3